### PR TITLE
add resolveDid

### LINKED ISSUE #4: Endpoint: 'resolveDid'
### Discussed in https://github.com/xmtp/didethresolver/discussions/3

<div type='discussions-op-text'>

<sup>Originally posted by **jac18281828** November 29, 2023</sup>
### Overview

The `resolveDid` endpoint is designed to receive an Ethereum public key and return did document in a JSON format. This endpoint is part of our decentralized data service, which provides access to decentralized identity.

### Endpoint

```
POST /api/v1/resolveDid
```

### Request Format

The request should be a JSON object containing one field: `publicKey`.

- `publicKey` (string, required): The Ethereum public key (starting with '0x').

Example Request:
```json
{
  "publicKey": "0x123abc..."
}
```

### Response Format

The response will be a JSON object containing a did document for related to the provided Ethereum public key. 

Example Response:
```json
{
  "@context": [
    "https://www.w3.org/ns/did/v1",
    "https://w3id.org/security/suites/ed25519-2020/v1"
  ],
  "controller": "did:ethr:0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad",
  "id": "did:ethr:0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad",
  "service": [
    {
      "id": "did:ethr:0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad",
      "recipientKeys": "0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad",
      "serviceEndpoint": "https://xmtp.com/resolver",
      "type": "MessagingService"
    }
  ],
  "verificationMethod": [
    {
      "controller": "did:ethr:0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad",
      "id": "did:ethr:0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad",
      "publicKeyMultibase": "0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad",
      "type": "Ed25519VerificationKey2020"
    }
  ]
```

### Error Handling

In case of an error (e.g., invalid public key, server error), the endpoint will return a JSON object with an `error` field describing the issue.

Example Error Response:
```json
{
  "error": "Invalid public key format"
}
```

### Security and Authentication

- The endpoint is open access.

### Future requirements
- Access control
- All requests must be made over HTTPS.
- Rate limiting is applied to prevent abuse.

### Support

Please refer to the DID specification: [DID](https://www.w3.org/TR/did-core/)</div>

### PR DESCRIPTION
**Pull request by insipx**
closes #4 

- [x] Add DID types (DIDDocument) and DID parsing (DIDurl)
- [x] resolve did history with the registry
- [x] wrap did document  
- [x] tests
- [x] update docs

Not implementing for this PR
- blockTag for DID

related to [roadmap#23](https://github.com/orgs/xmtp/projects/15/views/3?pane=issue&itemId=46353468)

# Example

Run against Sepolia with `RUST_LOG=didethresolver=debug ./target/release/didethresolver -p "wss://ethereum-sepolia.publicnode.com"`

example query: 
❯ `curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "did_resolveDid", "params": ["0x423B6F365C14F4b233928088FF937e01c143A328"]}' http://localhost:9944`

Pretty print the result:
❯ `curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "did_resolveDid", "params": ["0x423B6F365C14F4b233928088FF937e01c143A328"]}' http://localhost:9944 | jq .result`

### CODE REVIEW

File: src/types/did_url.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
+//! Convenience Wrapper around [`Url`] for DID URIs according to the [DID Spec](https://www.w3.org/TR/did-core/#did-syntax)
```
Comment (insipx) [inline src/types/did_url.rs:1]: This works well but the `url` crate which is based on the [whatwg standard](https://url.spec.whatwg.org/)

Ideally we find a crate that already implemented the standard DID is a derivative of, or we can just use regex like did-ethr reference implementation. It seems to me that the `url` crate is more flexible than regex, though.

**Review by jac18281828 — COMMENTED** [review]
This looks like a pretty good start!   I hit a few snags because the tests dont appear to build and there are a few warnings outstanding.

This project may be helpful:   https://github.com/OwnYourData/didlint

It appears there is a network api as well.

**Comment by insipx** [comment]
> This looks like a pretty good start! I hit a few snags because the tests dont appear to build and there are a few warnings outstanding.
> 
> This project may be helpful: https://github.com/OwnYourData/didlint
> 
> It appears there is a network api as well.

thanks! Definitely cleanup to do and still working on generating the actual did document from the did history. There is a network endpoint, `did_resolveDid` that can be called:

`curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "did_resolveDid", "params": ["0x6CEb0bF1f28ca4165d5C0A04f61DC733987eD6ad"]}' http://localhost:9944/ | jq .result`

but currently only returns a stub did document until `wrap_did_document` is finished.

**Comment by insipx** [comment]
This could use some more tests, but overall I feel like it's in a pretty good place. I'll create a ticket for the cases I'd like to create tests for

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+use anyhow::Result;
+use didethresolver::run;
+
+#[tokio::main]
+async fn main() -> Result<()> {
```
Comment (insipx) [inline src/main.rs:5]: This separation of a main and lib is done to make unit testing possible

**Review by tsachiherman — APPROVED** [review]

File: example.env [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ADDRESS="127.0.0.1:9944"
```
Comment (jac18281828) [inline example.env:1]: I believe these quotes are not required in this context.   I believe in the TypeScript dotenv these quotes will be included in the environment, not sure how rust behaves but its worth checking

File: src/lib.rs [hunk 4]
```diff
@@ -78,6 +78,79 @@
...
+
+const DEFAULT_ADDRESS: &str = "127.0.0.1:9944";
+const DEFAULT_PROVIDER: &str = "http://127.0.0.1:8545";
+
+// TODO: Get registry address from environment variable, or configuration file
+// in order to support multiple chains, we may need to support multiple providers via RPC
+// so it could be worth defining a config file that maps chainId to RPC provider (like
+// did-eth-resolver)
+/// The address of the DID Registry contract on the Ethereum Sepolia Testnet
+pub const DID_ETH_REGISTRY: &str = "0xd1D374DDE031075157fDb64536eF5cC13Ae75000";
```
Comment (jac18281828) [inline src/lib.rs:108]: This will be a proxy contract with a stable address on all networks.   Allowing override probably makes sense, but may also create a security loophole.

File: src/main.rs (hunk 2 above)
Comment (jac18281828) [reply src/main.rs:5]: Thank you!

File: src/resolver.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        log::debug!("Using deployed registry at {}", registry.address());
+        Ok(Self { signer, registry })
+    }
+
+    pub async fn resolve_did(&self, public_key: H160) -> Result<DidDocument> {
+        let history = self.changelog(public_key).await?;
+        Ok(self.wrap_did_document(public_key, history).await?)
+    }
+
+    async fn changelog(&self, public_key: H160) -> Result<Vec<(DIDRegistryEvents, LogMeta)>> {
```
Comment (jac18281828) [inline src/resolver.rs:45]: Are we handing revoked an expired events?    If that is not part of this change please make sure we open an issue on that before this gets merged.

**Review by jac18281828 — APPROVED** [review]
Thank you for adding tests and documentation!   This is great work and a great start.   Please add a ticket to support handling for revoked attributes as that is somewhat important

File: src/resolver.rs (hunk 5 above)
Comment (insipx) [reply src/resolver.rs:45]: it should handle these events in the `wrap_did_document` method, although we should definitely make sure of it and add tests specifically for revoked attributes, adde a ticket [here](https://github.com/xmtp/didethresolver/issues/22)

File: src/lib.rs (hunk 4 above)
Comment (insipx) [reply src/lib.rs:108]: we should be explicit that the override is only for tests, or restrict overriding the tests only when we're sure we're in a test environment
